### PR TITLE
#1666 organization units

### LINKED ISSUE #1666: Organization units for the Identity module

### PR DESCRIPTION
**Pull request by mperk**
Resolve #1666
Everything is done. If you can find a bug please send me. I want to fix.
I found a bug but I can't fix. I don't know why. I want to your help with this. You can see the bug: OrganizationUnitManager_Tests => AddRoleToOrganizationUnitAsync

### CODE REVIEW

**Comment by hikalkan** [comment]
Thank you @mperk I will review this.
@maliming can you also review it before me?
Thanks.

**Comment by maliming** [comment]
hi @mperk 
According to the current design, how do I get its users based on an ou id in the application service layer?

**Comment by mperk** [comment]
> hi @mperk
> According to the current design, how do I get its users based on an ou id in the application service layer?

hi @maliming 
I think I missed it. I will commit it as soon as. Thank you.

File: modules/identity/src/Volo.Abp.Identity.Domain/Volo/Abp/Identity/IdentityUserOrganizationUnit.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using Volo.Abp.Domain.Entities.Auditing;
+using Volo.Abp.MultiTenancy;
+
+namespace Volo.Abp.Identity
+{
+    /// <summary>
+    /// Represents membership of a User to an OU.
+    /// </summary>
+    public class IdentityUserOrganizationUnit : CreationAuditedEntity<Guid>, IMultiTenant, ISoftDelete
```
Comment (maliming) [inline modules/identity/src/Volo.Abp.Identity.Domain/Volo/Abp/Identity/IdentityUserOrganizationUnit.cs]: Can `IdentityUserOrganizationUnit` and `OrganizationUnitRole ` inherit from `Entity`?`

They do not need a primary key. Such as `IdentityUserRole`

**Comment by maliming** [comment]
```
class IdentityOrganizationUnit : AggregateRoot<Guid>
{

}

class IdentityUserOrganizationUnit : Entity
{
	Guid UserId
	
	Guid OrganizationUnitId
}

class IdentityRoleOrganizationUnit : Entity
{
	Guid RoleId
	
	Guid OrganizationUnitId
}


class IdentityUser : AggregateRoot<Guid>
{
	ICollection<IdentityUserOrganizationUnit> OrganizationUnits
}

class IdentityRole : AggregateRoot<Guid>
{
	ICollection<IdentityRoleOrganizationUnit> OrganizationUnits
}
```


`IOrganizationUnitRepository `is used to get its structure, users, roles. (`Getusers. Getroles`)

`IdentityUserManager `and `IdentityRoleManager `is used to manage the organization of users and roles. (`AddToOrg. RemoveOrg getOrgList` ..)

We can use `IOrganizationUnitRepository `in `IdentityUserStore `to get the role information of the user organization.

I'm not sure if we should merge `OrganizationUnitManager `into `IOrganizationUnitRepository`, because its methods all seem to be crud operations.

This is my simple idea, what do you think?

File: modules/identity/src/Volo.Abp.Identity.Domain/Volo/Abp/Identity/IdentityUserOrganizationUnit.cs (hunk 1 above)
Comment (mperk) [reply modules/identity/src/Volo.Abp.Identity.Domain/Volo/Abp/Identity/IdentityUserOrganizationUnit.cs]: You are right. I fixed. Thank you.

**Comment by mperk** [comment]
> ```
> class IdentityOrganizationUnit : AggregateRoot<Guid>
> {
> 
> }
> 
> class IdentityUserOrganizationUnit : Entity
> {
> 	Guid UserId
> 	
> 	Guid OrganizationUnitId
> }
> 
> class IdentityRoleOrganizationUnit : Entity
> {
> 	Guid RoleId
> 	
> 	Guid OrganizationUnitId
> }
> 
> 
> class IdentityUser : AggregateRoot<Guid>
> {
> 	ICollection<IdentityUserOrganizationUnit> OrganizationUnits
> }
> 
> class IdentityRole : AggregateRoot<Guid>
> {
> 	ICollection<IdentityRoleOrganizationUnit> OrganizationUnits
> }
> ```
> 
> `IOrganizationUnitRepository `is used to get its structure, users, roles. (`Getusers. Getroles`)
> 
> `IdentityUserManager `and `IdentityRoleManager `is used to manage the organization of users and roles. (`AddToOrg. RemoveOrg getOrgList` ..)
> 
> We can use `IOrganizationUnitRepository `in `IdentityUserStore `to get the role information of the user organization.
> 
> I'm not sure if we should merge `OrganizationUnitManager `into `IOrganizationUnitRepository`, because its methods all seem to be crud operations.
> 
> This is my simple idea, what do you think?

Me too, I am not sure but I think it should be seperate.

**Comment by crashsol** [comment]
In many cases, customers need to sort the organization by themselves. It is perfect to provide a sorting field in the basic module, 
`
public float sorting {get; set;}
`

Floating point can be easily sorted by dragging in the front end

**Comment by Nokecy** [comment]
any update?

**Comment by hikalkan** [comment]
I will deeply review this (probably) tomorrow.

**Comment by hikalkan** [comment]
Hi all,

Thank you @mperk for your great contribution. As overall, it is really fine. However, I found some problems with the implementation. Some of them are minor, some of them are related to design decisions, but some of them are bugs (for example, `OrganizationUnit` should not have a `Children` property which violates the aggregate rule: reference other aggregates only by id. We should remove this collection, otherwise it fails when we want to insert this entity to a document database like MongoDB).

So, I moved this to v2.2 milestone (we'll release v2.1 today), will review deeper and write my comments in details.

**Comment by hikalkan** [comment]
BTW, @mperk, it would be appreciated if you solve the conflicts.

**Comment by mperk** [comment]
> BTW, @mperk, it would be appreciated if you solve the conflicts.

I did it.

**Comment by vd3d** [comment]
Hi,

No news about this merge ? :-P

or is it for later (when ? next month ... ) ?

Thanks

**Comment by maliming** [comment]
@vd3d  It is planned in abp 2.2. March 05, 2020.

**Comment by hikalkan** [comment]
Thank you again @mperk for your great contribution. I checked it deeply and found some points to improve. However, I will handle them, finalize the feature and merge it in the v2.3 (sorry to move to to v2.3, but we need it). So, you don't need to make addinional commits.

Here, a list of my findings (notes for myself 🙂):

* IdentityUserCreateOrUpdateDtoBase.cs: OrganizationUnits -> OrganizationUnitIds (rename)
* Lozalication key `Identity.OrganizationUnit.MaxUserMembershipCount` is missing except English
* `Maximum length of the <see cref="Code"/> property` Remove Code reference from summary
* `throw new AbpException(string.Format("Can not set more than {0} organization unit for a user!", maxCount));`
  Should be business exception & localized.
* `IdentityUserOrganizationUnit` and `OrganizationUnitRole`: Cancel ISoftDelete implementation.
* Remove `public virtual ICollection<OrganizationUnit> Children { get; set; }` from `OrganizationUnit`. It won't work for MongoDB. We first should add test to prove it.
* `OrganizationUnitManager_Tests.AddRoleToOrganizationUnitAsync` is failing, should be fixed
* `tenantId` should be the last parameter of the `OrganizationUnit` and `OrganizationUnitRole` constructors, and it should be optional.
* IdentityDbContextModelBuilderExtensions.cs: Use `b.ConfigureByConvention();` for all entities.
* Remove `ConfigureAwait(false);` usages. They are not needed anymore.

**Comment by xewn** [comment]
这个功能，还不能合并吗？已经延后很多次了。正式版里面都还没有 organization units

**Comment by xewn** [comment]
这个PR不行，就用这个么。  https://gitee.com/xugege/KST.ABP.git

**Comment by zsanhong** [comment]
when will the organization module release? plan to release at  2.1 then delay to  2.2,then delay to 2.3,and now is 2.4 is comimg,will it delay again?

**Comment by maliming** [comment]
> will it delay again?

If you are eager for this feature, you can implement it yourself.

**Comment by xewn** [comment]
> > will it delay again?
> 
> If you are eager for this feature, you can implement it yourself.

收费版本内，没有组织结构，购买了的客户，如何正常的使用abp vNext？

**Comment by viewtance** [comment]
Look forward to see this feature early, thank you guys.

**Comment by hikalkan** [comment]
We are actively working on this. Hopefully it will be released with the v2.6 (April 20th planned date).

**Comment by xewn** [comment]
When is the Organization units for the Identity module  released?

**Comment by hikalkan** [comment]
Hopefully with this (2.8) release if we don't have any significiant problem. Still working on it.

**Comment by hikalkan** [comment]
Hi all,

Unfortunately, we won't be able to release this with v2.8. It is "almost" finished, but not well tested. This is a core feature (effects roles & permissions) and releasing it tomorrow (the release day) will be risky.

It will be definitely released with v2.9 (2 weeks later), sorry for the delay.

**Comment by mgochmuradov** [comment]
Hi @hikalkan, is there any documentation on how to use this feature?

**Comment by maliming** [comment]
hi @mgochmuradov 
You can temporarily refer to the explanation of aspnetboilerplate for this.
https://aspnetboilerplate.com/Pages/Documents/Zero/Organization-Units

**Comment by hikalkan** [comment]
Will be documented here: https://docs.abp.io/en/abp/latest/Modules/Identity

**Pull request merged**